fetch client endpoints are typed as `any` on Typescript `5.5.x`

### Describe the bug

When creating a client from my contract on a NodeJS server using Typescript `5.5.x` (tested with `.2`, `.3` and `.4`), the endpoints are typed as `any`, losing all type information.
My IDE reports that type and `tsc` as well.

![image](https://github.com/user-attachments/assets/580862f8-ce12-48da-af76-ed17e2fefa50)

It works fine on `5.4.5`

### How to reproduce

Edit: clone this repo: https://github.com/guillaume-docquier-vention/ts-rest-typescript-5.5 and follow the readme

- Setup a NodeJS project using Typescript `5.5.x`
- Create a contract
- Create a client from the contract, using `initClient`
- Use any endpoint from the client

Notice that the endpoints are not typed.

### Expected behavior

All endpoints should be typed.
or
The docs/package should say that it is not compatible with Typescript `5.5.x`

### Code reproduction

```ts
import { initContract, initClient } from "@ts-rest/core"
import { z } from "zod"

export const contract = initContract().router(
  {
    auth: {
      login: {
        method: "POST",
        path: `/auth/login`,
        body: z.object({
          email: z.string(),
          password: z.string(),
        }),
        responses: {
          200: z.object({
            token: z.string(),
          }),
        },
      },
    },
    logout: {
      method: "POST",
      path: `/logout`,
      body: z.object({
        email: z.string(),
        password: z.string(),
      }),
      responses: {
        200: z.object({
          token: z.string(),
        }),
      },
    },
  },
  {
    strictStatusCodes: true,
    pathPrefix: "/api",
  },
)

const client = initClient(contract, {
  baseUrl: `http://localhost:3000`,
  baseHeaders: {},
})

client.auth.login("invalid") // Problem: type of "login" is "any"
client.logout("invalid") // Problem: type of "logout" is "any"
client.invalid() // As expected: TS2339: Property invalid does not exist on type
```


### ts-rest version

3.46.0

### Other versions

Node 22.3.0
Npm 10.8.1
Typescript 5.5.2 / 5.5.3 / 5.5.4

I would say this is a duplicate of #618, and is an IDE error, not ts-rest or Typescript, but it's weird that `tsc` is not erroring for you.

You can see [here](https://www.typescriptlang.org/play/?strict=false&strictNullChecks=false&ts=5.5.3#code/JYWwDg9gTgLgBAbzsAdsGBhCKZQIYDGMANMmpgDbACmOcAvnAGZQQhwDkAAjAM4C0UarxgB6AtGocA3AChQkWIjgAvBs1bsOKiABMZs2dQAei+BJQi4F3IXgBeMuiw58RABQBKAHSsArjDUUO6ycIihYXB4AQAWAFzhkZEUEADmqAkIEUlhINQwMXoJHAAKAPIAygAqHMTZOWB4BQkABqLRBaIp6SgtdTmRAEZ6AJ4JKt4QgwBW1B5ZA0nUIHjAFOPeIlCoqV79i2GNvLwA7tC6G1s7e-VJ9J77A0K8kJbCmbc5AEwADD8bU1m80+AxgEAA1rRLrhrg8QZF7o8cvQkWEUfV0UluhAAh8BnkCkVOOVqrVPo1mnA2tiAn1PsNdGNVJMZnMYO4Fotlqt1syrihdnCDkdTudodsBTcBojPs9Xrx3olFr9-szAWyOfCwmDIShxbDUQihcikZiGPtOWErkQKjAmn5eFhdIrcH5qEiKTESkImMBjMV2mBgGS0XVPHJZBYrAQqLQHE5KDQcO4bG4SErBngFQBVKC8loxGAwMBxURdCAEPAUQoiOIAZj+PzpYUzCoAEtQ8M6oLxMuj7hGY0mYN4OjFvN1UO4OKgAG5V4D6cOR2M4CdpHHsmcoedUJdyIdx7xzhe6LxyIA), and in the output below that that the types are evaluated correctly, and erroring with invalid input.

```shell
$ tsc -v
Version 5.5.4
$ tsc
src/test.ts:46:19 - error TS2559: Type '"invalid"' has no properties in common with type '{ cache?: RequestCache; body?: { email?: string; password?: string; }; fetchOptions?: FetchOptions; extraHeaders?: Record<string, string>; overrideClientOptions?: Partial<...>; }'.

46 client.auth.login('invalid');
                     ~~~~~~~~~

src/test.ts:47:15 - error TS2559: Type '"invalid"' has no properties in common with type '{ cache?: RequestCache; body?: { email?: string; password?: string; }; fetchOptions?: FetchOptions; extraHeaders?: Record<string, string>; overrideClientOptions?: Partial<...>; }'.

47 client.logout('invalid');
                 ~~~~~~~~~

src/test.ts:48:8 - error TS2339: Property 'invalid' does not exist on type 'RecursiveProxyObj<RecursivelyApplyOptions<{ auth: { login: { method: "POST"; path: "/auth/login"; body: ZodObject<{ email: ZodString; password: ZodString; }, "strip", ZodTypeAny, { email?: string; password?: string; }, { ...; }>; responses: { ...; }; }; }; logout: { ...; }; }, { ...; }>, { ...; }>'.

48 client.invalid();
          ~~~~~~~
```

Please submit a link to a full repo with this problem.

Disregarding the `tsc` issue. IDEs will very often take a long time to compute or infer complex types, so while this is happening in the background the type will show as `any` and you would not get any errors, red line, or completion until that happens. Even after the types are determined, complex types might still show as `any` 

I'm curious, because this might be related to your other random autocompletion issues. Do you have this "[Use types from server](https://www.jetbrains.com/help/webstorm/typescript-support.html#ws_ts_restart_ts_service)" option turned on in WebStorm? I tried it and noticed it was a bit unstable and very often super slow or even fails to compute types sometimes.

--- a/.changeset/green-pots-laugh.md
+++ b/.changeset/green-pots-laugh.md
@@ -1,0 +1,5 @@
+---
+'@ts-rest/core': patch
+---
+
+Fix broken types on TS 5.5 in Node environments without `lib.dom`

--- a/libs/ts-rest/core/package.json
+++ b/libs/ts-rest/core/package.json
@@ -19,9 +19,13 @@
   },
   "license": "MIT",
   "peerDependencies": {
+    "@types/node": "^18.18.7 || >=20.8.4",
     "zod": "^3.22.3"
   },
   "peerDependenciesMeta": {
+    "@types/node": {
+      "optional": true
+    },
     "zod": {
       "optional": true
     }

--- a/libs/ts-rest/core/src/lib/client.spec.ts
+++ b/libs/ts-rest/core/src/lib/client.spec.ts
@@ -212,7 +212,7 @@ type ClientGetPostsType = Expect<
         } & Record<string, string | undefined>;
         fetchOptions?: FetchOptions;
         overrideClientOptions?: Partial<OverrideableClientArgs>;
-        cache?: RequestCache;
+        cache?: FetchOptions['cache'];
       }
     | undefined
   >
@@ -241,7 +241,7 @@ type ClientWithoutBaseHeadersGetPostsType = Expect<
       } & Record<string, string | undefined>;
       fetchOptions?: FetchOptions;
       overrideClientOptions?: Partial<OverrideableClientArgs>;
-      cache?: RequestCache;
+      cache?: FetchOptions['cache'];
     }
   >
 >;
@@ -265,7 +265,7 @@ type ClientGetPostType = Expect<
       } & Record<string, string | undefined>;
       fetchOptions?: FetchOptions;
       overrideClientOptions?: Partial<OverrideableClientArgs>;
-      cache?: RequestCache;
+      cache?: FetchOptions['cache'];
     }
   >
 >;
@@ -773,7 +773,7 @@ describe('client', () => {
         baseUrl: 'http://localhost:5002',
       });
 
-      global.fetch = jest.fn(() =>
+      globalThis.fetch = jest.fn(() =>
         Promise.resolve({
           json: () =>
             Promise.resolve({
@@ -797,7 +797,7 @@ describe('client', () => {
         } as RequestInit,
       });
 
-      expect(global.fetch).toHaveBeenCalledWith(
+      expect(globalThis.fetch).toHaveBeenCalledWith(
         'http://localhost:5002/posts/1',
         {
           cache: undefined,
@@ -812,7 +812,7 @@ describe('client', () => {
           },
         },
       );
-      (global.fetch as jest.Mock).mockClear();
+      (globalThis.fetch as jest.Mock).mockClear();
     });
   });
 });
@@ -866,7 +866,7 @@ type CustomClientGetPostsType = Expect<
       } & Record<string, string | undefined>;
       fetchOptions?: FetchOptions;
       overrideClientOptions?: Partial<OverrideableClientArgs>;
-      cache?: RequestCache;
+      cache?: FetchOptions['cache'];
       uploadProgress?: (progress: number) => void;
     }
   >
@@ -891,7 +891,7 @@ type CustomClientGetPostType = Expect<
       } & Record<string, string | undefined>;
       fetchOptions?: FetchOptions;
       overrideClientOptions?: Partial<OverrideableClientArgs>;
-      cache?: RequestCache;
+      cache?: FetchOptions['cache'];
       uploadProgress?: (progress: number) => void;
     }
   >

--- a/libs/ts-rest/core/src/lib/client.ts
+++ b/libs/ts-rest/core/src/lib/client.ts
@@ -42,11 +42,18 @@ export type AppRouteFunction<
   ? (args?: Prettify<TArgs>) => Promise<Prettify<ClientInferResponses<TRoute>>>
   : (args: Prettify<TArgs>) => Promise<Prettify<ClientInferResponses<TRoute>>>;
 
-export type FetchOptions = Omit<RequestInit, 'method' | 'headers' | 'body'>;
+export type FetchOptions = typeof globalThis extends {
+  Request: infer T extends typeof Request;
+}
+  ? Omit<
+      NonNullable<ConstructorParameters<T>[1]>,
+      'method' | 'headers' | 'body'
+    >
+  : never;
 
 export interface OverrideableClientArgs {
   baseUrl: string;
-  credentials?: RequestCredentials;
+  credentials?: FetchOptions['credentials'];
   jsonQuery?: boolean;
   validateResponse?: boolean;
 }
@@ -56,35 +63,38 @@ export interface ClientArgs extends OverrideableClientArgs {
   api?: ApiFetcher;
 }
 
-export type ApiFetcherArgs = {
-  route: AppRoute;
-  path: string;
-  method: string;
-  headers: Record<string, string>;
-  body: FormData | URLSearchParams | string | null | undefined;
-  rawBody: unknown;
-  rawQuery: unknown;
-  contentType: AppRouteMutation['contentType'];
-  fetchOptions?: FetchOptions;
-  validateResponse?: boolean;
+export type ApiFetcherArgs<TFetchOptions extends FetchOptions = FetchOptions> =
+  {
+    route: AppRoute;
+    path: string;
+    method: string;
+    headers: Record<string, string>;
+    body: FormData | URLSearchParams | string | null | undefined;
+    rawBody: unknown;
+    rawQuery: unknown;
+    contentType: AppRouteMutation['contentType'];
+    fetchOptions?: FetchOptions;
+    validateResponse?: boolean;
 
-  /**
-   * @deprecated Use `fetchOptions.credentials` instead
-   */
-  credentials?: RequestCredentials;
-  /**
-   * @deprecated Use `fetchOptions.signal` instead
-   */
-  signal?: AbortSignal;
-  /**
-   * @deprecated Use `fetchOptions.cache` instead
-   */
-  cache?: RequestCache;
-  /**
-   * @deprecated Use `fetchOptions.next` instead
-   */
-  next?: { revalidate?: number | false; tags?: string[] } | undefined;
-};
+    /**
+     * @deprecated Use `fetchOptions.credentials` instead
+     */
+    credentials?: TFetchOptions['credentials'];
+    /**
+     * @deprecated Use `fetchOptions.signal` instead
+     */
+    signal?: TFetchOptions['signal'];
+    /**
+     * @deprecated Use `fetchOptions.cache` instead
+     */
+    cache?: 'cache' extends keyof TFetchOptions
+      ? TFetchOptions['cache']
+      : never;
+    /**
+     * @deprecated Use `fetchOptions.next` instead
+     */
+    next?: 'next' extends keyof TFetchOptions ? TFetchOptions['next'] : never;
+  };
 
 export type ApiFetcher = (args: ApiFetcherArgs) => Promise<{
   status: number;
@@ -249,7 +259,7 @@ export const fetchApi = (options: FetchApiOptions) => {
     ...(fetchOptions?.cache && { cache: fetchOptions.cache }),
     ...(fetchOptions &&
       'next' in fetchOptions &&
-      !!fetchOptions?.next && { next: fetchOptions.next }),
+      !!fetchOptions?.next && { next: fetchOptions.next as any }),
   };
 
   if (route.method !== 'GET') {

--- a/libs/ts-rest/core/src/lib/infer-types.spec.ts
+++ b/libs/ts-rest/core/src/lib/infer-types.spec.ts
@@ -553,7 +553,7 @@ it('type inference helpers', () => {
           } & Record<string, string | undefined>;
           fetchOptions?: FetchOptions;
           overrideClientOptions?: Partial<OverrideableClientArgs>;
-          cache?: RequestCache;
+          cache?: FetchOptions['cache'];
         };
         createPost: {
           body: { title: string; content: string };
@@ -564,7 +564,7 @@ it('type inference helpers', () => {
           } & Record<string, string | undefined>;
           fetchOptions?: FetchOptions;
           overrideClientOptions?: Partial<OverrideableClientArgs>;
-          cache?: RequestCache;
+          cache?: FetchOptions['cache'];
         };
         uploadImage: {
           body:
@@ -580,7 +580,7 @@ it('type inference helpers', () => {
           } & Record<string, string | undefined>;
           fetchOptions?: FetchOptions;
           overrideClientOptions?: Partial<OverrideableClientArgs>;
-          cache?: RequestCache;
+          cache?: FetchOptions['cache'];
         };
         nested: {
           getComments: {
@@ -597,7 +597,7 @@ it('type inference helpers', () => {
             } & Record<string, string | undefined>;
             fetchOptions?: FetchOptions;
             overrideClientOptions?: Partial<OverrideableClientArgs>;
-            cache?: RequestCache;
+            cache?: FetchOptions['cache'];
           };
         };
       }
@@ -614,7 +614,7 @@ it('type inference helpers', () => {
           extraHeaders?: Record<string, string | undefined>;
           fetchOptions?: FetchOptions;
           overrideClientOptions?: Partial<OverrideableClientArgs>;
-          cache?: RequestCache;
+          cache?: FetchOptions['cache'];
         };
       }
     >

--- a/libs/ts-rest/core/src/lib/infer-types.ts
+++ b/libs/ts-rest/core/src/lib/infer-types.ts
@@ -242,7 +242,9 @@ type ClientInferRequestBase<
       /**
        * @deprecated Use `fetchOptions.cache` instead
        */
-      cache?: RequestCache;
+      cache?: 'cache' extends keyof TFetchOptions
+        ? TFetchOptions['cache']
+        : never;
 
       /**
        * @deprecated Use `fetchOptions.next` instead

--- a/libs/ts-rest/core/tsconfig.spec.json
+++ b/libs/ts-rest/core/tsconfig.spec.json
@@ -3,7 +3,7 @@
   "compilerOptions": {
     "outDir": "../../../dist/out-tsc",
     "module": "commonjs",
-    "types": ["jest", "node"]
+    "types": ["jest"]
   },
   "include": ["jest.config.ts", "**/*.test.ts", "**/*.spec.ts", "**/*.d.ts"]
 }

--- a/libs/ts-rest/next/src/lib/next-client.spec.ts
+++ b/libs/ts-rest/next/src/lib/next-client.spec.ts
@@ -39,7 +39,7 @@ describe('next-client', () => {
       extraHeaders?: Test['extraHeaders'];
       fetchOptions?: FetchOptions;
       overrideClientOptions?: Partial<OverrideableClientArgs>;
-      cache?: RequestCache;
+      cache?: FetchOptions['cache'];
       next?: FetchOptions['next'];
     };
     type NextClientTypeTest = Expect<Equal<Test, ExpectedClientArgs>>;

